### PR TITLE
Store offset commit metadata when calling `rd_kafka_offsets_store`

### LINKED ISSUE #3927: `rd_kafka_offsets_store` does not store and send offset commit metadata
Description
===========

Offset commit metadata can be used to send additional data to the broker when making an offset commit. See this [doc from the Java consumer](https://kafka.apache.org/21/javadoc/org/apache/kafka/clients/consumer/OffsetAndMetadata.html#:~:text=The%20Kafka%20offset%20commit%20API,the%20commit%20was%20made%2C%20etc.).

Librdkafka currently sends the offset commit metadata when the commit is requested by calling `rd_kafka_commit_queue` but not when calling `rd_kafka_offsets_store` to use the auto-commit.

How to reproduce
================

1. Using `confluent-kafka-go`, which supports offset commit metadata, store an offset with metadata to be auto-committed:
```go
package main

import (
	"fmt"

	"github.com/confluentinc/confluent-kafka-go/kafka"
)

func main() {
	consumer, err := kafka.NewConsumer(&kafka.ConfigMap{
		"bootstrap.servers":        "localhost:9092",
		"group.id":                 "test",
		"enable.auto.offset.store": false,
		"enable.auto.commit":       true,
	})
	if err != nil {
		panic(err)
	}
	_, version := kafka.LibraryVersion()
	fmt.Printf("Started consumer version:%s\n", version)

	err = consumer.Subscribe("test", func(consumer *kafka.Consumer, event kafka.Event) error {
		fmt.Printf("%v\n", event.String())
		return nil
	})
	if err != nil {
		panic(err)
	}

	for {
		event := consumer.Poll(1000)
		if event == nil {
			fmt.Println("No event")
			continue
		}
		switch e := event.(type) {
		case kafka.Error:
			fmt.Printf("%s\n", e.Error())
		case *kafka.Message:
			fmt.Printf("%s: %s\n", e.TopicPartition, e.Value)
			metadata := "test metadata"
			_, err = consumer.StoreOffsets([]kafka.TopicPartition{{
				Topic:     e.TopicPartition.Topic,
				Partition: e.TopicPartition.Partition,
				Offset:    e.TopicPartition.Offset + 1,
				Metadata:  &metadata,
			}})
			if err != nil {
				panic(err)
			}
		}
		break
	}
	err = consumer.Close()
	if err != nil {
		panic(err)
	}
}
```

```
Started consumer version:1.9.1
AssignedPartitions: [test[0]@unset]
No event
test[0]@1: hello
RevokedPartitions: [test[0]@unset]
```

2. Using [`kafka-python`](https://kafka-python.readthedocs.io/en/master/), print the offsets for that consumer group:
```python
from kafka import KafkaAdminClient

config = {"bootstrap_servers": "localhost:9092"}
client = KafkaAdminClient(**config)
print(client.list_consumer_group_offsets("test"))
```

```python
{TopicPartition(topic='test', partition=0): OffsetAndMetadata(offset=1, metadata='')}
```

3. Now replace `StoreOffsets` with `CommitOffsets` in the above go script to make the commit synchronously and run the same python script:

```
Started consumer version:1.9.1
AssignedPartitions: [test[0]@unset]
No event
test[0]@2: hello
RevokedPartitions: [test[0]@unset]
```

```python
{TopicPartition(topic='test', partition=0): OffsetAndMetadata(offset=2, metadata='test metadata')}
```

Investigation
===========

[`rd_kafka_offset_store`](https://github.com/edenhill/librdkafka/blob/v1.8.2/src/rdkafka_offset.c#L647:L665) should pass the metadata to [`rd_kafka_offset_store0`](https://github.com/edenhill/librdkafka/blob/52949110d60e36c9eae59d31956a7fd760d783fa/src/rdkafka_offset.h#L84:L104) so that it can store the last metadata sent in the topic partition `toppar`.

It can then be recovered when re-creating the offsets list from stored data [here](https://github.com/edenhill/librdkafka/blob/4faeb8132521da70b6bcde14423a14eb7ed5c55e/src/rdkafka_partition.c#L3273).

https://github.com/edenhill/librdkafka/compare/master...mathispesch:librdkafka:store-offsets-metadata is a proposed patch, that I can continue working on to try and factorise it and add tests.

Checklist
=========

 - [x] librdkafka version (release number or git tag): `1.9.1`
 - [x] Apache Kafka version: `2.7.2`
 - [x] librdkafka client configuration: `enable.auto.commit=true, enable.auto.offset.store=false`
 - [x] Operating system: `Ubuntu 20.04`
 - [x] Provide logs (with `debug=..` as necessary) from librdkafka: https://gist.github.com/mathispesch/1df4d26873fa483b202497a198936c76
 - [x] Provide broker log excerpts: no relevant logs, see output of kafka admin calls instead
 - [ ] Critical issue

### PR DESCRIPTION
**Pull request by mathispesch**
Fixes #3927.

Offset commit metadata can be used to send additional data to the broker when making an offset commit. See this [doc from the Java consumer](https://kafka.apache.org/21/javadoc/org/apache/kafka/clients/consumer/OffsetAndMetadata.html#:~:text=The%20Kafka%20offset%20commit%20API,the%20commit%20was%20made%2C%20etc.).

Librdkafka currently sends the offset commit metadata when the commit is requested by calling `rd_kafka_commit_queue` but not when calling `rd_kafka_offsets_store` to use the auto-commit. This PR fixes that by storing the offset commit metadata when calling `rd_kafka_offsets_store`.

### CODE REVIEW

**Comment by mathispesch** [comment]
Hey @edenhill,
Could you please give this PR a look? :slightly_smiling_face:

**Comment by mathispesch** [comment]
Hey @edenhill or @ijuma,
Could you please give this PR a look? 🙂

**Comment by ijuma** [comment]
cc @emasab

**Comment by mathispesch** [comment]
Hey folks,
Have you had the chance to look into this PR?

**Review by milindl — COMMENTED** [review]
Looks good to me, besides the tiny issue with the test. I'd like @emasab to weigh in.
We should add back the CHANGELOG, but under the next version, (we can use 2.1.0, or just call it vNext or something).

File: tests/0130-store_offsets.c [hunk 1]
```diff
@@ -108,6 +117,44 @@ static void do_test_store_unassigned(void) {
                     "offset %" PRId64 ", not %" PRId64,
                     proper_offset, rkmessage->offset);
 
+        TEST_SAY(
+            "Retrieving committed offsets to verify committed offset "
+            "metadata\n");
+        rd_kafka_topic_partition_list_t *committed_toppar;
+        committed_toppar = rd_kafka_topic_partition_list_new(1);
```
[the remark this answers is not in the document]
Comment (mathispesch) [reply tests/0130-store_offsets.c:124]: Good catch! Fixed this in 5d37ab1afdb6c81c64691d97430d59e754348946.

**Comment by emasab** [comment]
@mathispesch Thanks for the contribution! For security reason currently it's not allowed to run SemaphoreCI on forked PRs. I merge it into an internal branch and open a new PR. If you want to do additional changes before merging to master please open a new PR based on that branch.

**Comment by mensfeld** [comment]
@mathispesch @emasab should it also store metadata when using `rd_kafka_offsets_store` and then `rd_kafka_commit` with `null` or not? :thinking:

**Pull request merged**